FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RecipientsViewModel.kt

```diff
@@ -0,0 +1,391 @@
+
+            if (cachedRecipientWithPubKeys == null) {
+              roomDatabase.recipientDao().insertSuspend(RecipientEntity(email = emailLowerCase))
+              cachedRecipientWithPubKeys =
+                roomDatabase.recipientDao().getRecipientWithPubKeysByEmailSuspend(emailLowerCase)
+            } else {
+              for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+                try {
+                  val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+                  publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+                } catch (e: Exception) {
+                  e.printStackTrace()
+                  publicKeyEntity.isNotUsable = true
+                }
+              }
+            }
+
+            try {
+              getPgpInfoFromServer(email = emailLowerCase)?.let { pgpKeyDetailsList ->
+                cachedRecipientWithPubKeys?.let { recipientWithPubKeys ->
+                  updateCachedInfoWithPubKeysFromLookUp(
+                    recipientWithPubKeys,
+                    pgpKeyDetailsList
+                  )
+                }
+              }
+              cachedRecipientWithPubKeys = getCachedRecipientWithPubKeys(emailLowerCase)
+              cachedRecipientWithPubKeys?.let { recipients.add(it) }
+            } catch (e: Exception) {
+              e.printStackTrace()
+              ExceptionUtil.handleError(e)
+            }
+          }
+        }
+        setResultForRemoteContactsLiveData(type, Result.success(recipients))
+      } catch (e: Exception) {
+        e.printStackTrace()
+        ExceptionUtil.handleError(e)
+        setResultForRemoteContactsLiveData(type, Result.exception(e))
+      }
+    }
+  }
+
+  private suspend fun getCachedRecipientWithPubKeys(emailLowerCase: String): RecipientWithPubKeys? =
+    withContext(Dispatchers.IO) {
+      val cachedRecipientWithPubKeys = roomDatabase.recipientDao()
+        .getRecipientWithPubKeysByEmailSuspend(emailLowerCase) ?: return@withContext null
+
+      for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+        try {
+          val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+          publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+        } catch (e: Exception) {
+          e.printStackTrace()
+          publicKeyEntity.isNotUsable = true
+        }
+      }
+      return@withContext cachedRecipientWithPubKeys
+    }
+
+  private suspend fun updateCachedInfoWithPubKeysFromLookUp(
+    cachedRecipientEntity: RecipientWithPubKeys, list: List<PgpKeyDetails>
+  ) = withContext(Dispatchers.IO) {
+    val email = cachedRecipientEntity.recipient.email
+    val mapOfUniquePubKeysFromLookUp = mutableMapOf<String, PgpKeyDetails>()
+
+    for (pgpKeyDetails in list) {
+      val fingerprint = pgpKeyDetails.fingerprint
+      val existingPgpKeyDetails = mapOfUniquePubKeysFromLookUp[fingerprint]
```

Comment:
Please call this `alreadyEncounteredFetchedPgpKeyDetails` because here the `existing` alone seemed like you're referring to local database, which it's not. That way, it will be clearer.